File: src/main/java/sample/AccountRepository.java
package sample;

import org.springframework.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
class AccountRepository {

    final private DatabaseClient databaseClient;

    public AccountRepository(DatabaseClient databaseClient) {
        this.databaseClient = databaseClient;
    }

    Mono<Account> findById(Long id) {
        String q = "SELECT balance FROM accounts WHERE id = :id";
        return databaseClient.sql(q)
                .bind("id", id)
                .map(row -> new Account(id, row.get("balance", Long.class)))
                .one();
    }

    Mono<Integer> charge(Long id, Long amount) {
        String q = "UPDATE accounts SET balance = balance - :amount WHERE id = :id";
        return databaseClient.sql(q)
                .bind("id", id)
                .bind("amount", amount)
                .fetch()
                .rowsUpdated();
    }

}

File: src/main/kotlin/sample/AccountRepository.kt
package sample

import org.springframework.data.r2dbc.core.DatabaseClient
import org.springframework.data.r2dbc.core.awaitRowsUpdated
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class AccountRepository(
  private val databaseClient: DatabaseClient
) {

  fun findById(id: Long): Mono<Account> {
    val q = "SELECT balance FROM accounts WHERE id = :id"
    return databaseClient.execute(q)
      .bind("id", id)
      .map { row ->
        Account(
          id = id,
          balance = row.getLong("balance")!!
        )
      }
      .one()
  }

  fun updateBalance(contractId: Long, balance: Long): Mono<Int> {
    val q = "UPDATE accounts SET balance = :balance WHERE id = :id"
    return databaseClient.execute(q)
      .bind("id", contractId)
      .bind("balance", balance)
      .fetch()
      .rowsUpdated()
  }

  suspend fun save(account: Account): Int {
    val insert =
      "INSERT INTO accounts (id, balance) VALUES ($1, $2)"
    return databaseClient.execute(insert)
        .bind(0, account.id)
        .bind(1, account.balance)
        .fetch()
        .awaitRowsUpdated()
  }

}